routers/user.controller.js

```diff
@@ -0,0 +1,245 @@
+//     }
+//   };
+
+//   static userLogin = async (req, res, next) => {
+//     try {
+//       const { email, password } = req.body;
+
+//       const user = await userModel.findOne({ email });
+
+//       if (!user) {
+//         return res.status(401).send({ message: "Email or password is wrong" });
+//       }
+
+//       console.log("password", password);
+//       console.log("user.passwordHash", user.passwordHash);
+
+//       const isPassValid = await bcrypt.compare(password, user.passwordHash);
+
+//       if (isPassValid) {
+//         return res.status(402).send({ message: "Email or password is wrong" });
+//       }
+
+//       const token = await user.generateAndSaveToken();
+
+//       await userModel.findByIdAndUpdate(user._id, {
+//         token,
+//       });
+
+//       res.status(200).send({
+//         user: {
+//           id: user._id,
+//           name: user.username,
+//           email: user.email,
+//           familyId: user.familyId
+//         },
+//         activeToken: token,
+//       });
+//     } catch (err) {
+//       next(err);
+//     }
+//   };
+// }
+
+// module.exports = userController;
+const Joi = require("joi");
+const bcrypt = require("bcrypt");
+const UserModel = require("../database/modules/user.model");
+const {
+  ApiError,
+  errorHandler,
+  validate,
+  createVerificationToken,
+  mailer,
+} = require("../helpers");
+const authorization = require("../middlewares/auth");
+
+// const authRouter = Router();
+class userController {
+  constructor() {}
+  async userRegister(req, res) {
+    try {
+      validate(
+        Joi.object({
+          name: Joi.string(),
+          email: Joi.string().email().required(),
+          password: Joi.string().min(4),
+        }),
+        req.body
+      );
+
+      const { name, email, password } = req.body;
+
+      const [user] = await UserModel.find({ email });
+
+      if (user) throw new ApiError(409, "Email are in use");
+
+      console.log(typeof createVerificationToken);
+
+      const verificationToken = createVerificationToken();
+
+      await UserModel.create({
+        name,
+        email,
+        passwordHash: password,
+        verificationToken,
+      });
+
+      await mailer.sendMail(email, verificationToken);
+
+      res.status(201).send({ name, email });
+    } catch (err) {
+      errorHandler(req, res, err);
+    }
+  }
+
+  //==================================================================
+
+  async userLogin(req, res) {
+    try {
+      validate(
+        Joi.object({
+          email: Joi.string().email().required(),
+          password: Joi.string().min(4).required(),
+        }),
+        req.body
+      );
+
+      const { email, password } = req.body;
+
+      const foundUser = await UserModel.findOne({ email });
+
+      if (!foundUser) {
+        throw new ApiError(401, "Email or password is wrong");
+      }
+      if (foundUser.verificationToken) {
+        throw new ApiError(428, "Email not verified");
```

Comment:
If you would like to use precondition status code here, I think 412 would suite that error more